alephium/contracts/sequence/sequence.ral

```diff
@@ -0,0 +1,62 @@
+// sequence started from 0, set `next = 0` when init contract state
+Abstract Contract Sequence(
+    mut next: U256,
+    mut next1: U256, // TODO: better name
+    mut next2: U256,
+    undoneSequenceTemplateId: ByteVec,
+    refundAddress: Address
+) extends Constants() {
+    @using(readonly = true)
+    fn setExecuted(offset: U256, current: U256) -> U256 {
+        let flag = 1 << offset
+        assert!(current & flag == 0, ErrorCodes.InvalidSequence)
+        return current | flag
+    }
+
+    @using(assetsInContract = true)
+    fn compact() -> () {
+        if (next1 != SequenceExecutedFlag) {
+            let encodedState = encodeToByteVec!(
+                selfContractId!(),
+                next,
+                next1,
+                refundAddress
+            )
+            copyCreateSubContract!{selfAddress!() -> 1 alph}(u256To8Byte!(next / 256), undoneSequenceTemplateId, encodedState)
```

Comment:
Question: Would it be possible that an attacker from the other chain flood the chain with small transfer messages (small amount), drains the ALPH balance from the `Sequence` and  therefore disable it? I guess someone would then need to call the `deposit` method to top up the balance?

Reply:
Yes, if this happens we need to remove these `UndoneSequence` contracts by the `DestroyUndoneSequenceContracts` governance action.

Reply:
Would that make the legitimate transfers during the attack invalid? Tokens are deposited correctly on the other side, but we wont be able to withdraw it from ALPH.

Reply:
The `UndoneSequence` contract is used to store those messages that have not been withdrawn in time so that the sequence contract will not be blocked, if the normal user withdraws in time, then the `UndoneSequence` only have the sequences from attack messages. So it depends on when we destroy these undone sequence contracts.